File: vm/src/Vm.cpp
#include "../include/Vm.h"
#include <stack>
#include <stdexcept>
#include <string>

void Vm::Load(const std::vector<Instruction>& code)
{
	instructions = code;
}

int Vm::Run()
{
	std::stack<int> stack;
	size_t ip = 0;

	while (ip < instructions.size())
	{
		const Instruction& instr = instructions[ip];
		switch (instr.Op)
		{
		case Opcode::Const:
			stack.push(instr.Value);
			ip++;
			break;
		case Opcode::Return:
			if (stack.empty())
				return 0;
			{
				int value = stack.top();
				stack.pop();
				return value;
			}
		default:
			throw std::runtime_error("Неизвестная инструкция по адресу " + std::to_string(ip));
		}
	}

	return stack.empty() ? 0 : stack.top();
}

File: packages/vm/src/Vm.ts
import { Instruction, Opcode } from './types'

class Vm
{
	private instructions: Instruction[] = []

	load(code: Instruction[]): void
	{
		this.instructions = code
	}

	run(): number
	{
		const stack: number[] = []
		let ip = 0

		while (ip < this.instructions.length)
		{
			const instr = this.instructions[ip]

			switch (instr.op)
			{
				case Opcode.Const:
					stack.push(instr.value)
					ip++
					break

				case Opcode.Return:
					return stack.length > 0 ? stack[stack.length - 1] : 0

				default:
					throw new Error(`Неизвестная инструкция по адресу ${ip}`)
			}
		}

		return stack.length > 0 ? stack[stack.length - 1] : 0
	}
}

export {
	Vm,
}
